### ISSUE TITLE
Strange behaviour of percentile iterator

### ISSUE THREAD

**Issue by algermissen** [jonhoo/hdrsample#66]
I am trying to create hgrm text output. For this I am using this function: https://gist.github.com/algermissen/9f86fe6051a5e4f89ef92d8f5cfd9637 which aims to be a port of https://github.com/HdrHistogram/HdrHistogram_py/blob/5b9b94a1827a88c4782100dff52dfc7e2578ee87/hdrh/histogram.py#L576

The histogram is created in Rust code with `let hist = Histogram::<u32>::new_with_bounds(1, 60 * 60 * 1000, 2)`, exported as V2+DEFLATE and turned to base64.

The hgrm outputs below are 1. Generated from the Rust histogram using the function in the gist and 2. created by pasting the base64 into https://hdrhistogram.github.io/HdrHistogramJSDemo/decoding-demo.html

What strikes me as odd (besides that they are not the same) is that the Rust version seems to be 'one-off' in some sense. It is as if the last iteration should not be there (the percentile 1.0 appears twice. In the gist I check `if v.quantile() < 1.0` and the iteration value seems to indeed return 1.0 twice.

````
      951.00 1.000000000000         10
      951.00 1.000000000000         10
````

Is this an indicator of a bug or am I doing something very wrong here?

I also wonder why there is not better precision for the percentile column shown in the Rust output - is this my formatting or some unwanted rounding going on (note: I am still new to Rust, so might entirely be me, being simply stupid)

Rust-direct-output with Gist-Function:
````
       Value     Percentile TotalCount 1/(1-Percentile)

       79.00 0.100000000000          1           1.11
       79.00 0.100000000000          1           1.11
       85.00 0.300000000000          3           1.43
       89.00 0.400000000000          4           1.67
       89.00 0.400000000000          4           1.67
       98.00 0.500000000000          5           2.00
      115.00 0.600000000000          6           2.50
      133.00 0.700000000000          7           3.33
      133.00 0.700000000000          7           3.33
      212.00 0.800000000000          8           5.00
      212.00 0.800000000000          8           5.00
      212.00 0.800000000000          8           5.00
      373.00 0.900000000000          9          10.00
      373.00 0.900000000000          9          10.00
      373.00 0.900000000000          9          10.00
      373.00 0.900000000000          9          10.00
      373.00 0.900000000000          9          10.00
      951.00 1.000000000000         10
      951.00 1.000000000000         10
#[Mean    =       221.90, StdDeviation   =       257.55]
#[Max      =          951, TotalCount   =           10]
#[Buckets      =           15, SubBuckets   =         2048]
````

Converted from base64 using JS-demo page:
````
       Value     Percentile TotalCount 1/(1-Percentile)

       79.00 0.000000000000          1           1.00
       79.00 0.100000000000          1           1.11
       85.00 0.200000000000          3           1.25
       85.00 0.300000000000          3           1.43
       89.00 0.400000000000          4           1.67
       98.00 0.500000000000          5           2.00
      115.00 0.550000000000          6           2.22
      115.00 0.600000000000          6           2.50
      133.00 0.650000000000          7           2.86
      133.00 0.700000000000          7           3.33
      212.00 0.750000000000          8           4.00
      212.00 0.775000000000          8           4.44
      212.00 0.800000000000          8           5.00
      373.00 0.825000000000          9           5.71
      373.00 0.850000000000          9           6.67
      373.00 0.875000000000          9           8.00
      373.00 0.887500000000          9           8.89
      373.00 0.900000000000          9          10.00
      951.00 0.912500000000         10          11.43
      951.00 1.000000000000         10
#[Mean    =       221.90, StdDeviation   =       257.55]
#[Max     =       951.00, Total count    =           10]
#[Buckets =           15, SubBuckets     =          256]
````

**Comment by marshallpierce** [comment]
I'll take a look. In the meantime, note that we recently changed some floating-point subtleties of how percentiles are calculated to what I think is the most mathematically precise (and floating-point stable) way of doing it: https://github.com/jonhoo/hdrsample/pull/58/files. The key change is in the [`value_at_quantile` function in `lib.rs`](https://github.com/jonhoo/hdrsample/blob/master/src/lib.rs#L1232); rounding now behaves a different way. The TypeScript port uses a different rounding method: https://github.com/HdrHistogram/HdrHistogramJS/blob/master/src/AbstractHistogram.ts#L403 For more context on that, read back a long way in the [Gitter room](https://gitter.im/HdrHistogram/HdrHistogram). 

So, in summary, it's not surprising that they might be different, but the repetition looks a little odd, so I'll fiddle with it.

**Comment by marshallpierce** [comment]
I've added a quantile-iteration example based on your gist to the `print-quantiles` branch: https://github.com/jonhoo/hdrsample/compare/print-quantiles?expand=1.

I don't think the use of the histogram's significant figures to format the value is appropriate, though the Java impl does it (and the Python one does as well to mirror the Java one, presumably). Some precision was lost as values/counts were recorded, depending on significant figures chosen when creating the histogram, but it does no good IMO to further muddy the waters by re-applying the significant figures to what's already been mangled. I'd rather see the values exactly as they are in the histogram, because that's going to be the most accurate: if my original value of `10_000_000` has become `10_002_431` because of loss of precision, I'd like to know that that's exactly what's going on. So, in the version I made, the value is presented exactly as-is.

Also, I've made the quantile display precision configurable via `--quantile-precision`.

Here's an example invocation to play with. It writes 1000 numbers between 0 and 10 million into a histogram, then outputs it:

```
jot 1000 0 10000000 | cg run --example cli -- serialize -c | cg run --example cli -- quantiles -t 5
```

(`cg` is my alias for `cargo` because who has time for all those characters)

**Comment by jonhoo** [comment]
What is `jot` above? Is `seq` not what you want?

**Comment by marshallpierce** [comment]
`jot` is the same sort of thing as `seq`. IIRC, `seq` is the GNU not-invented-here clone of BSD's `jot`, or something like that.

**Comment by marshallpierce** [comment]
I think that the `more` implementation for the quantile iterator is behaving poorly with the new quantile calculations. I suspect that that "go one more step" was to make the old quantile rounding show 100%, and now it's going past 100%. Adding some hopefully self-explanatory printlns shows that forcing the iterator to go one extra step heads into mathematical nonsense and a duplicate final line:

```
     9994239 0.9989999999999999991118215802998747676611        999        1000.00
quantile to iterate to 0.9988281250000001776356839400250464677811, incremented by 0.0001953125, halvings 9
     9994239 0.9989999999999999991118215802998747676611        999        1000.00
quantile to iterate to 0.9990234375000002220446049250313080847263, incremented by 0.0001953125, halvings 9
     9994239 0.9989999999999999991118215802998747676611        999        1000.00
quantile to iterate to 0.9991210937500002442490654175344388931990, incremented by 0.00009765625, halvings 10
    10002431 1.0000000000000000000000000000000000000000       1000              ∞
Reached total count
quantile more
quantile to iterate to 1.1000000000000000888178419700125232338905, incremented by 0.1, halvings 0
    10002431 1.0000000000000000000000000000000000000000       1000              ∞
```

**Comment by marshallpierce** [comment]
I'm assuming that #67 resolved this; feel free to re-open if that's not the case.
